### PR TITLE
Add pre-commit hooks as Github actions

### PR DESCRIPTION
**Pull request by ulgens**

### CODE REVIEW

**Review by hartwork — CHANGES_REQUESTED** [review]
Hi @ulgens this approach does not seem to scale very well: Whenever we add a new check to the pre-commit config we'd need to extend CI as well. And I don't really see value in having the checks all split up, do you? Can't we run all pre-commit checks at a single place at once? I'm doing that in other projects and red or green and a bit of output was all I ever needed from it. What do you think?

**Comment by ulgens** [comment]
> Can't we run all pre-commit checks at a single place at once? I'm doing that in other projects and red or green and a bit of output was all I ever needed from it.

Can you please give an example of this? If that's a possibility, I'd like to use it but as far as I know, pre-commit doesn't allow to run of all hooks at once with `--all-files` parameter.

**Comment by hartwork** [comment]
> Can you please give an example of this? If that's a possibility, I'd like to use it but as far as I know, pre-commit doesn't allow to run of all hooks at once with `--all-files` parameter.

@ulgens I'm not using the original GitHub action at https://github.com/hartwork/jawanndenn/blob/23221ab1db852d8eb07ac1074719530acd9c6f8a/.github/workflows/run_pre_commit.yml#L34 but that's one way to do it. It doesn't have to be this very way.

**Comment by hartwork** [comment]
@ulgens PS: With jawanndenn locally I see this:

```console
# pre-commit run --all-files
pyupgrade....................................................................Passed
Check JSON...............................................(no files to check)Skipped
Check Yaml...................................................................Passed
Check for merge conflicts....................................................Passed
Fix End of Files.............................................................Passed
Trim Trailing Whitespace.....................................................Passed
flake8.......................................................................Passed
isort........................................................................Passed
Check .rst files with the same linter used by pypi.org.......................Passed
```

**Comment by ulgens** [comment]
Oh okay. I remember there was a discussion on pre-commit repo about running all hooks at once and it closed as unresolved, but I think they solved it later. Not testing it is my bad.

File: .github/workflows/pre-commit.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# https://github.com/pre-commit/action
+
+name: "pre-commit"
+
+on:
+  push:
+    branches:
+      - master
+  pull_request: ~
```
Comment (hartwork) [inline .github/workflows/pre-commit.yml:9]: What is the intention of tilde (`~`) here? If it's intentional, we maybe need a comment.

File: .github/workflows/pre-commit.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# https://github.com/pre-commit/action
```
Comment (hartwork) [inline .github/workflows/pre-commit.yml:1]: Would you mind sharing that code under the MIT license and adding an MIT license header here? Mimicking the format at… https://github.com/hahey/django-migration-vis/blob/c97016d764571cad269ab0ed3a8c926dfed25256/django_migration_vis/management/commands/visualizemigrations.py#L1-L3 …would be perfect.

**Review by hartwork — CHANGES_REQUESTED** [review]
@ulgens please see two minor details below.

File: .github/workflows/pre-commit.yml (hunk 1 above)
Comment (ulgens) [reply .github/workflows/pre-commit.yml:9]: That means an empty list and it's part of markdown syntax. Not sure what comment is needed.

File: .github/workflows/pre-commit.yml (hunk 2 above)
Comment (ulgens) [reply .github/workflows/pre-commit.yml:1]: That project is not mine and i don't intend to maintain it with a different license.

File: .github/workflows/pre-commit.yml (hunk 1 above)
Comment (hartwork) [reply .github/workflows/pre-commit.yml:9]: Interesting! Let's drop it then, lone `pull_request:` works just fine.

Comment (hartwork) [reply .github/workflows/pre-commit.yml:9]: Turns out it's `None`, not an empty list. In Python:

```
In [1]: import yaml

In [2]: yaml.safe_load('foo: ~')
Out[2]: {'foo': None}

In [3]: yaml.safe_load('foo:')
Out[3]: {'foo': None}

In [4]: yaml.safe_load("foo: []")
Out[4]: {'foo': []}
```

And in the Yaml spec at https://yaml.org/spec/1.2-old/spec.html#id2805071 .

Comment (hartwork) [reply .github/workflows/pre-commit.yml:9]: @ulgens actually let's also drop the limit on `master` branch. We want CI everywhere. Plain…

```yaml
on:
  push:
  pull_request:
```

…will be perfect, if we add cron later.

File: .github/workflows/pre-commit.yml (hunk 2 above)
Comment (hartwork) [reply .github/workflows/pre-commit.yml:1]: I think we may have a misunderstanding. Let's try to clear it up. The project is licensed under MIT right now. It is good practice to have a copyright header mentioning the license in all files possible. And by adding that line you'll confirm to the community (not to keep maintaining this file or project but) that you yourself had the right to contribute this patch (e.g. some work contracts would take away those rights from you and then you cannot contribute anywhere) and agree to the MIT license. What do you think?

Comment (ulgens) [reply .github/workflows/pre-commit.yml:1]: Oh okay got it, thanks for the clarification.

File: .github/workflows/pre-commit.yml (hunk 1 above)
Comment (ulgens) [reply .github/workflows/pre-commit.yml:9]: I'll check it again. The problem I was trying to solve there is, "on push, pull_request" config triggers all tests twice on the pull request with no visible benefit.

Comment (hartwork) [reply .github/workflows/pre-commit.yml:9]: @ulgens it only looks like running the same thing twice, it's actually two different things: One CI is testing the commit at the head of the pull request, the other is doing a temporary merge commit and tests the merged result. Often there's little difference, semantically one is testing the pull request in isolation and the other if merging will break the target branch. We want both.

Comment (hartwork) [reply .github/workflows/pre-commit.yml:9]: @ulgens for proof:

In pull request https://github.com/hartwork/jawanndenn/pull/411:
- The branch HEAD is at commit 4e9a6e7468e7564cf7018643cba2b14dbe064d24
- One CI run `Build and test using Docker` is testing that commit (see https://github.com/hartwork/jawanndenn/runs/5218188939?check_suite_focus=true#step:2:364)
- The other CI run `Build and test using Docker` is testing the merge result (see https://github.com/hartwork/jawanndenn/runs/5218189320?check_suite_focus=true#step:2:361)

Comment (hartwork) [reply .github/workflows/pre-commit.yml:9]: @ulgens maybe I was too quick here and my reply probably came across in a non-ideal way. I now understand the idea better and I see that it's a tradeoff, either too much CI or too little, either potentially waiting and more CO2 or more chance to miss some subtle breakage. Does it matter to you which approach we pick here? I'm more open to giving the current proposal a try now, if you want.

Comment (ulgens) [reply .github/workflows/pre-commit.yml:9]: @hartwork About push vs pull request, I don't think it matters on a repository that had no commits in the last 2.5 years. My intention with all of these commits is to recover the repo from dead and make it useful again, in the fastest way possible. Forking it was an option, but I didn't want to deal with submitting an orphan package on PyPI. That being said, top-notch code quality is always important but I don't think fixing an extra empty space on a commit or discussions about CI trigger config provides any extra value.

That being said, I tend to not having (seemingly) duplicate Cı checks but I'm okay with whatever you see fit.

File: .github/workflows/pre-commit.yml (hunk 2 above)
Comment (ulgens) [reply .github/workflows/pre-commit.yml:1]: @hartwork On second thought, and again I'm okay with whatever you see fit but I'm just sharing my opinion, I believe license information belongs to the root directory of repo, not as duplicated lines in each file (and it creates clutter) Github detect top-level LICENSE files and assign them to repo attributes.

Comment (hartwork) [reply .github/workflows/pre-commit.yml:1]: @ulgens the problem starts e.g. when people want to copy files to other projects, when you have mixed license code bases and so on. GitHub's detection cannot be relied on, they only show one single license and do not even distinct between GPL-2 and GPL-2+ properly. Do they by now? I have seen situations where files without explicit headers were a problem and the FSF also explicitly recommends it:

> Give each file the proper copyright notices. Make sure to clearly identify which versions of the license users can use.

From https://www.gnu.org/licenses/gpl-howto.en.html . I see no reason to limit that to GPL.

File: .github/workflows/pre-commit.yml (hunk 1 above)
Comment (hartwork) [reply .github/workflows/pre-commit.yml:9]: Hi @ulgens ,

> My intention with all of these commits is to recover the repo from dead and make it useful again, in the fastest way possible.

Personally, I want quality over speed. Why go as fast as possible? If my review is too picky, I can offer we flip this around i.e. you open bugs, I create PRs, and you get to take the review seat, instead. What do you think?

> Forking it was an option, but I didn't want to deal with submitting an orphan package on PyPI.

I'm not sure how orphan packages on PyPI work. Could you elaborate or share a link to what you refer to?

> That being said, top-notch code quality is always important but I don't think fixing an extra empty space on a commit or discussions about CI trigger config provides any extra value.

I think can agree to disagree here. Stray whitespace changes in commits are something that goes into Git history forever, and it's a time waster and annoying when inspecting past history, and it's cheap to fix and unnecessary in my  view. (I have a blog post on [my approach to code review](https://blog.hartwork.org/posts/my-approach-to-code-review/) if you're curious.)

> I tend to not have (seemingly) duplicate CI checks but I'm okay with whatever you see fit.

Okay, I would vote for the simpler but more covering version then, i.e.
```
on:
  push:
  pull_request:
```

File: .github/workflows/pre-commit.yml (hunk 2 above)
Comment (ulgens) [reply .github/workflows/pre-commit.yml:5]: Fixed.

File: .github/workflows/pre-commit.yml (hunk 1 above)
Comment (ulgens) [reply .github/workflows/pre-commit.yml]: Fixed.

**Pull request merged**